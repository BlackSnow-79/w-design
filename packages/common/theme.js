"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var hex2Rgba_1 = require("../utils/hex2Rgba");
var colors_1 = require("./colors");
// breakpoint values
exports.breakpoints = [24, 32, 48, 64, 80].map(function (n) { return n + "em"; });
exports.fonts = {
    0: "system-ui, sans-serif",
    sans: "-apple-system, BlinkMacSystemFont, system-ui, sans-serif",
    base: 'Roboto, "Helvetica Neue"'
};
exports.fontSizes = [12, 14, 16, 20, 24, 32, 48, 64, 72, 96];
exports.fontWeights = {
    normal: 400,
    bold: 700,
    xbold: 900
};
// for any scale, either array or objects will work
exports.lineHeights = [1, 1.125, 1.25, 1.5];
exports.letterSpacings = {
    normal: "normal",
    caps: "0.25em"
};
// space is used for margin and padding scales
// it's recommended to use powers of two to ensure alignment
// when used in nested elements
// numbers are converted to px
exports.space = [0, 4, 8, 16, 32, 64, 128];
exports.radii = [0, 2, 4];
exports.shadows = [
    "none",
    "0 .2rem .5rem " + hex2Rgba_1.default(colors_1.colors.N50, 0.5),
    "inset 0 0 0 1px rgba(87, 85, 217, .2), 0 0 4px rgba(87, 85, 217, .2)"
];
exports.default = {
    fonts: exports.fonts,
    fontSizes: exports.fontSizes,
    fontWeights: exports.fontWeights,
    lineHeights: exports.lineHeights,
    colors: colors_1.colors,
    space: exports.space,
    breakpoints: exports.breakpoints,
    letterSpacings: exports.letterSpacings
};
//# sourceMappingURL=theme.js.map