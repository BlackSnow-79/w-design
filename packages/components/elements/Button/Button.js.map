{"version":3,"file":"Button.js","sourceRoot":"","sources":["../../../../src/components/elements/Button/Button.tsx"],"names":[],"mappings":";;;;;;;;;;;;;;;;;;;;;;;;AAAA,6BAA+B;AAC/B,mDAAkD;AAClD,mDAAkD;AAClD,gDAA2C;AAG3C,IAAM,YAAY,GAAG,gBAAM,CAAC,MAAM,gFAAA,MAC9B,EAAc,KACjB,KADG,+BAAc,CACjB,CAAC;AACF,YAAY,CAAC,WAAW,GAAG,cAAc,CAAC;AAE1C,IAAM,CAAC,GAAG,gBAAM,CAAC,CAAC,iGAAA,kBAEZ,EAAc,UAEnB,KAFK,+BAAc,CAEnB,CAAC;AACF,CAAC,CAAC,WAAW,GAAG,SAAS,CAAC;AAE1B,IAAM,IAAI,GAAG,gBAAM,CAAC,IAAI,gFAAA,MACpB,EAAc,KACjB,KADG,+BAAc,CACjB,CAAC;AACF,IAAI,CAAC,WAAW,GAAG,YAAY,CAAC;AAEhC;IAAqB,0BAAyB;IAA9C;QAAA,qEAuEC;QA/CC,uCAAuC;QACvC,aAAO,GAAG;YACR,KAAI,CAAC,KAAK,CAAC,UAAU,CAAC,CAAC,CAAC,SAAS,CAAC,CAAC,CAAC,KAAI,CAAC,KAAK,CAAC,OAAO,CAAC;QACzD,CAAC,CAAC;QAEF,YAAM,GAAG,UAAC,CAAoC;YAC5C,IAAI,KAAI,CAAC,KAAK,CAAC,MAAM,EAAE;gBACrB,KAAI,CAAC,KAAK,CAAC,MAAM,CAAC,CAAC,CAAC,CAAC;aACtB;QACH,CAAC,CAAC;QAEF,aAAO,GAAG,UAAC,CAAoC;YAC7C,IAAI,KAAI,CAAC,KAAK,CAAC,OAAO,EAAE;gBACtB,KAAI,CAAC,KAAK,CAAC,OAAO,CAAC,CAAC,CAAC,CAAC;aACvB;QACH,CAAC,CAAC;QAEF,kBAAY,GAAG,UAAC,GAAgB;YAC9B,KAAI,CAAC,MAAM,GAAG,GAAG,CAAC;YAClB,IAAI,KAAI,CAAC,KAAK,CAAC,SAAS;gBAAE,KAAI,CAAC,KAAK,CAAC,SAAS,CAAC,GAAG,CAAC,CAAC;QACtD,CAAC,CAAC;;IA2BJ,CAAC;IAzDC,kCAAiB,GAAjB;QACE,IAAI,IAAI,CAAC,KAAK,CAAC,SAAS,IAAI,IAAI,CAAC,MAAM,EAAE;YACvC,IAAI,CAAC,MAAM,CAAC,KAAK,EAAE,CAAC;SACrB;IACH,CAAC;IAED,qCAAoB,GAApB;QACE,IAAI,CAAC,MAAM,GAAG,IAAI,CAAC;IACrB,CAAC;IAwBD,6BAAY,GAAZ;QACE,IAAI,IAAI,CAAC,KAAK,CAAC,IAAI,EAAE;YACnB,OAAO,IAAI,CAAC,KAAK,CAAC,UAAU,CAAC,CAAC,CAAC,IAAI,CAAC,CAAC,CAAC,CAAC,CAAC;SACzC;QAED,OAAO,YAAY,CAAC;IACtB,CAAC;IAED,uBAAM,GAAN;QACQ,IAAA,eAAqC,EAAnC,sBAAQ,EAAE,0BAAU,CAAgB;QAC5C,IAAM,WAAW,GAAG,+BAAc,CAAC,IAAW,CAAC,CAAC;QAEhD,IAAM,eAAe,GAAG,IAAI,CAAC,YAAY,EAAE,CAAC;QAE5C,OAAO,CACL,oBAAC,eAAe,aACd,GAAG,EAAE,IAAI,CAAC,YAAY,EACtB,OAAO,EAAE,IAAI,CAAC,OAAO,EACrB,UAAU,EAAE,UAAU,IAClB,WAAkB,GAErB,QAAQ,CACO,CACnB,CAAC;IACJ,CAAC;IAnEM,mBAAY,GAAG;QACpB,UAAU,EAAE,SAAS;QACrB,IAAI,EAAE,QAAQ;QACd,WAAW,EAAE,IAAI;QACjB,SAAS,EAAE,KAAK;QAChB,UAAU,EAAE,KAAK;QACjB,UAAU,EAAE,KAAK;KAClB,CAAC;IAEK,kBAAW,GAAG,iBAAiB,CAAC;IA2DzC,aAAC;CAAA,AAvED,CAAqB,KAAK,CAAC,SAAS,GAuEnC;AAED,kBAAe,MAAM,CAAC"}