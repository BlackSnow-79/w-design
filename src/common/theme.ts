import { colors } from "./colors";

export const fonts = {
  0: "system-ui, sans-serif",
  sans:
    '-apple-system, BlinkMacSystemFont, Roboto, "Helvetica Neue", sans-serif, system-ui, sans-serif'
};

export default {
  fonts,
  colors
};
