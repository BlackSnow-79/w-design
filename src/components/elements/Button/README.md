```js
<Provider>
  <Button>Default button</Button>{" "}
  <Button btnType="primary">Primary button</Button>{" "}
  <Button btnType="danger">Danger button</Button>{" "}
  <Button btnType="warning">Warning button</Button>{" "}
  <Button btnType="success">Success button</Button>{" "}
  <Button href="#" btnType="link">
    Link button
  </Button>{" "}
</Provider>
```

```js
<Provider>
  <Button borderRadius={4}>Default button</Button>{" "}
</Provider>
```
