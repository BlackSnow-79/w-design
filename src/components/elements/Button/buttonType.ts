export type BtnProps = {
  /** button apperance type */
  appearance?:
    | "default"
    | "primary"
    | "danger"
    | "warning"
    | "success"
    | "link";
  /** button apperance size */
  btnSize?: "sm" | "md" | "lg";
  /** fit container button*/
  fluid?: boolean;
  /** html button aria-controls*/
  ariaControls?: string;
  /** html button aria-expanded*/
  ariaExpanded?: boolean;
  /** html button aria-label*/
  ariaLabel?: string;
  /** html button aria-haspopup*/
  ariaHaspopup?: boolean;
  /** custom className*/
  className?: string;
  /** id of the button*/
  id?: string;
  /** button ref*/
  buttonRef?: React.Ref<any>;
  /** children */
  children?: React.ReactChild | string;
  /** button link url */
  href?: string;
  /** passing target attribute, if href available */
  target?: string;
  /** left icon*/
  iconLeft?: boolean;
  /** right icon*/
  iconRight?: boolean;
  /** button type attribute*/
  type?: "button" | "submit" | "reset";
  /** Name property of a linked form that the button submits when clicked. */
  form?: string;
  /** tabIndex assigning*/
  tabIndex?: number;
  /** loading state*/
  isLoading?: boolean;
  /** selected button */
  isSelected?: boolean;
  /** apply disabled button */
  isDisabled?: boolean;
};
