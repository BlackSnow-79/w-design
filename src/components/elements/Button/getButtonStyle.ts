import { css } from "../../../utils/styled";
import theme from "../../../common/theme";
import hex2Rgba from "../../../utils/hex2Rgba";
import { SpinnerStyle } from "../Spinner/Spinner";

const { colors } = theme;

/**
 * getAppearanceProps
 * @param appearance take the apperance props
 * @returns style applies to each button appearance
 */
const getAppearanceProps = (
  appearance?: "default" | "primary" | "danger" | "warning" | "success" | "link"
) => {
  const appearances = {
    default() {
      return `
        border: 0.05rem solid ${colors.N3};
        background: ${colors.N3};
        color: ${colors.N90};

        &:hover {
          background: ${colors.N4};
          border-color: ${colors.N4};
        }

        &:focus, &:active {
          background: ${colors.B5};
          border-color: ${colors.B5};
        }

        &:focus {
          color: ${colors.B60};
        }

        &:active {
          color: ${colors.N90};
        }
      `;
    },
    primary() {
      return `
        border: 0.05rem solid ${colors.B60};
        background: ${colors.B50};
        color: ${colors.N1};

        &:hover {
          background: ${colors.B40};
        }

        &:focus {
          background: ${colors.B40};
          border-color: ${colors.B60};
          box-shadow: 0 0 0 0.1rem ${hex2Rgba(colors.B5)};
          color: ${colors.WHITE};
        }

        &:active {
          background: ${colors.B60};
          border-color: ${colors.B70};
          color: ${colors.N1};
        }
      `;
    },
    danger() {
      return `
        border: 0.05rem solid ${colors.R40};
        background: ${colors.R30};
        color: ${colors.N1};

        &:hover {
          background: ${colors.R20};
        }

        &:focus {
          background: ${colors.R30};
          border-color: ${colors.R40};
          box-shadow: 0 0 0 0.1rem ${hex2Rgba(colors.R5)};
          color: ${colors.WHITE};
        }

        &:active {
          background: ${colors.R40};
          border-color: ${colors.R50};
          color: ${colors.N1};
        }
      `;
    },
    warning() {
      return `
        border: 0.05rem solid ${colors.O50};
        background: ${colors.O40};
        color: ${colors.N1};

        &:hover {
          background: ${colors.O30};
        }

        &:focus {
          background: ${colors.O30};
          border-color: ${colors.O40};
          box-shadow: 0 0 0 0.1rem ${hex2Rgba(colors.O10)};
          color: ${colors.WHITE};
        }

        &:active {
          background: ${colors.O40};
          border-color: ${colors.O50};
          color: ${colors.N1};
        }
      `;
    },
    success() {
      return `
        border: 0.05rem solid ${colors.G40};
        background: ${colors.G30};
        color: ${colors.N1};

        &:hover {
          background: ${colors.G20};
        }

        &:focus {
          background: ${colors.G30};
          border-color: ${colors.G40};
          box-shadow: 0 0 0 0.1rem ${hex2Rgba(colors.G7)};
          color: ${colors.WHITE};
        }

        &:active {
          background: ${colors.G40};
          border-color: ${colors.G50};
          color: ${colors.N1};
        }
      `;
    },
    link() {
      return `
        color: ${colors.B50};
        border: none;
        &:hover {
          background: ${colors.N3};
          color: ${colors.B60};
        }
        &:focus {
          background: ${colors.B5};
        }
        &:active {
          background: ${colors.B5};
          box-shadow: 0 0 0 0.1rem ${hex2Rgba(colors.B50)};
        }
      `;
    }
  };

  return appearances[appearance || "default"];
};

/**
 * getSizeProps
 * Take each size props as params to return style
 */
const getSizeProps = (props?: "sm" | "md" | "lg") => {
  const sizes = {
    sm() {
      return `
        font-size: .7rem;
        height: 1.4rem;
        padding: .15rem .3rem;
      `;
    },
    md() {
      return `
        height: 1.8rem;
        font-size: 0.8rem;
        padding: 0.35rem 0.4rem;
      `;
    },
    lg() {
      return `
        font-size: .9rem;
        height: 2rem;
        padding: .45rem .6rem;
      `;
    }
  };

  return sizes[props || "md"];
};

/**
 * getLoadingState
 * Apply style to each appearance when loading props is true
 */
const getLoadingState = (props: any) => {
  if (props.isLoading) {
    if (props.appearance === "default" || props.appearance === "link") {
      return SpinnerStyle;
    }

    return css`
      ${SpinnerStyle};
      &::after {
        border-color: ${colors.WHITE};
        border-right-color: transparent;
        border-top-color: transparent;
      }
    `;
  }

  return;
};

export const getButtonStyle = (props: any) => {
  /** Button type style*/
  const buttonAppearance = getAppearanceProps(props.appearance);

  /** Variable styled*/
  const buttonSize = getSizeProps(props.elementSize);

  let width;
  if (props.fluid) width = "100%";

  const loadingStyle = getLoadingState(props);

  return css`
    cursor: pointer;
    outline: 0;
    border-radius: 0.1rem;
    display: inline-block;
    text-decoration: none;
    text-align: center;
    transition: all 0.2s ease;
    user-select: none;
    vertical-align: middle;
    white-space: nowrap;
    width: ${width};
    &:disabled,
    &[disabled] {
      cursor: not-allowed;
      color: ${colors.N7} !important;
      background: ${colors.N4} !important;
      border: none;
    }

    ${buttonSize};
    ${buttonAppearance};
    ${loadingStyle};
  `;
};
