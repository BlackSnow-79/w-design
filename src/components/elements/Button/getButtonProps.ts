import { BtnProps } from "./buttonType";

const getAppearanceProps = (props: BtnProps) => {
  const {
    appearance,
    elementSize,
    fluid,
    className,
    isDisabled,
    isLoading
  } = props;

  return {
    appearance,
    elementSize,
    fluid,
    className,
    disabled: isDisabled,
    isLoading
  };
};

const getLinkElementProps = (props: BtnProps) => {
  const { href, target } = props;
  return { href, target };
};

const getButtonElementProps = (props: BtnProps) => {
  const { ariaHaspopup, ariaExpanded, ariaControls, form, type } = props;

  return {
    "aria-haspopup": ariaHaspopup,
    "aria-expanded": ariaExpanded,
    "aria-controls": ariaControls,
    form,
    type
  };
};

export const getButtonProps = (component: any) => {
  const { props } = component;

  const defaultProps = {
    ...getAppearanceProps(props),
    id: props.id,
    "aria-label": props.ariaLabel
  };

  if (props.href) {
    if (props.isDisabled) return defaultProps;

    return {
      ...defaultProps,
      ...getLinkElementProps(props)
    };
  }

  return {
    ...defaultProps,
    ...getButtonElementProps(props)
  };
};
