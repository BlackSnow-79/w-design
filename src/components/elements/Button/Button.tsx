import * as React from "react";
import { getButtonStyle } from "./getButtonStyle";
import { getButtonProps } from "./getButtonProps";
import styled from "../../../utils/styled";
import { BtnProps } from "./buttonType";

const StyledButton = styled.button`
  ${getButtonStyle};
`;
StyledButton.displayName = "StyledButton";

const A = styled.a`
  ${getButtonStyle};
`;
A.displayName = "StyledA";

const Span = styled.span`
  ${getButtonStyle};
`;
Span.displayName = "StyledSpan";

class Button extends React.Component<BtnProps> {
  static defaultProps = {
    appearance: "default",
    type: "button",
    btnSize: "md",
    isDisabled: false,
    isSelected: false
  };

  getComponent() {
    if (this.props.href) return this.props.isDisabled ? Span : A;

    return StyledButton;
  }

  render() {
    const { buttonRef, children } = this.props;
    const buttonProps = getButtonProps(this as any);

    const StyledComponent = this.getComponent();

    return (
      <StyledComponent ref={buttonRef} {...buttonProps as any}>
        {children}
      </StyledComponent>
    );
  }
}

export default Button;
