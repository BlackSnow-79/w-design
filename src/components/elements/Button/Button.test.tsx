import * as React from "react";
import { shallow, mount } from "enzyme";
import "jest-styled-components";
import Button from "./Button";
import theme from "../../../common/theme";

const { colors } = theme;

describe("<Button/>", () => {
  // Render test
  it("should renders correctly", () => {
    const wrapper = shallow(<Button>Test</Button>);
    expect(wrapper).toMatchSnapshot();
  });

  // Tag render test
  it("render as a link element", () => {
    const wrapper = mount(<Button href="#" appearance="link" />);
    expect(wrapper.find("a").length).toBe(1);
  });

  // Test render style
  it("render right style", () => {
    const wrapper = mount(<Button appearance="primary" />);
    expect(wrapper).toHaveStyleRule("background", `${colors.B50}`);
  });

  // Child test
  it("should render Child component", () => {
    const wrapper = mount(
      <Button>
        <div>Child</div>
      </Button>
    );
    expect(
      wrapper.contains(
        <Button>
          <div>Child</div>
        </Button>
      )
    ).toBeTruthy();
  });

  // Props test
  it("should has className", () => {
    const wrapper = shallow(<Button className="test" />);
    expect(wrapper.prop("className")).toEqual("test");
  });
});
