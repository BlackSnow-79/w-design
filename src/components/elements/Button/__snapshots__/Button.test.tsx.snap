// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button/> should renders correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Button
    appearance="default"
    btnSize="md"
    isDisabled={false}
    isLoading={false}
    isSelected={false}
    type="button"
  >
    Test
  </Button>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "appearance": "default",
      "aria-controls": undefined,
      "aria-expanded": undefined,
      "aria-haspopup": undefined,
      "aria-label": undefined,
      "children": "Test",
      "className": undefined,
      "disabled": false,
      "elementSize": undefined,
      "fluid": undefined,
      "form": undefined,
      "id": undefined,
      "isLoading": false,
      "type": "button",
    },
    "ref": null,
    "rendered": "Test",
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "appearance": "default",
        "aria-controls": undefined,
        "aria-expanded": undefined,
        "aria-haspopup": undefined,
        "aria-label": undefined,
        "children": "Test",
        "className": undefined,
        "disabled": false,
        "elementSize": undefined,
        "fluid": undefined,
        "form": undefined,
        "id": undefined,
        "isLoading": false,
        "type": "button",
      },
      "ref": null,
      "rendered": "Test",
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
